### PR TITLE
Upmaster

### PR DESCRIPTION
**Pull request by fsx950223**
The PR fix iou loss and ema bugs.
Add dataloader test case.

### CODE REVIEW

File: efficientdet/keras/efficientdet_keras.py [hunk 1]
```diff
@@ -728,13 +728,27 @@ def __init__(self, model_name=None, config=None, name=''):
...
       self._name = super().__init__(name, zero_based)
 
+  def build(self, input_shape):
+    super().build(input_shape)
+    ema_vars = self.trainable_variables
+    for v in self.variables:
+      # We maintain mva for batch norm moving mean and variance as well.
+      if 'moving_mean' in v.name or 'moving_variance' in v.name:
+        ema_vars.append(v)
+    self.ema_vars = ema_vars
```
Comment (mingxingtan) [inline efficientdet/keras/efficientdet_keras.py:750]: Where are ema vars created in eager mode?

File: efficientdet/keras/train_lib.py [hunk 2]
```diff
@@ -430,14 +441,28 @@ class and box losses from all levels.
             box_loss_layer([num_positives_sum, box_targets_at_level],
                            box_outputs[level]))
 
-      if self.config.iou_loss_type and self.loss.get('box_iou_loss', None):
-        box_iou_loss_layer = self.loss['box_iou_loss']
-        box_iou_losses.append(
-            box_iou_loss_layer([num_positives_sum, box_targets_at_level],
-                               box_outputs[level]))
+    if self.config.iou_loss_type:
+      input_anchors = BoxList(
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib.py:445]: Don't you want to wrap them as a function or a loss layer?

File: efficientdet/iou_utils.py [hunk 3]
```diff
@@ -42,7 +42,13 @@ def _grad_v(dv):
       gdh = -dv * 8 * arctan * width / (math.pi**2)
       return [gdh, gdw]
 
-    return v, _grad_v
+    def _grad_v_graph(dv, variables):
+      gdw = dv * 8 * arctan * height / (math.pi**2)
+      gdh = -dv * 8 * arctan * width / (math.pi**2)
+      return [gdh, gdw], tf.gradients(v, variables, grad_ys=dv)
+
+    return v, _grad_v if tf.compat.v1.executing_eagerly_outside_functions(
```
[the remark this answers is not in the document]
Comment (fsx950223) [reply efficientdet/iou_utils.py:50]: Fix https://github.com/google/automl/issues/583.
I have to do this to unify custom_gradient behavior.

File: efficientdet/keras/efficientdet_keras.py (hunk 1 above)
Comment (fsx950223) [reply efficientdet/keras/efficientdet_keras.py:750]: They will be created in train_step which is wrapped with tf.function.
The code only record variables that need to create ema.

File: efficientdet/keras/train_lib.py (hunk 2 above)
Comment (fsx950223) [reply efficientdet/keras/train_lib.py:445]: Good, I do it later.

File: efficientdet/keras/efficientdet_keras.py (hunk 1 above)
Comment (mingxingtan) [reply efficientdet/keras/efficientdet_keras.py:750]: sg

**Comment by mingxingtan** [comment]
Overall looks good to merge. Since it changes the dataloader and det_model_fn, I will  first verify it works for the original training pipeline before merging.

**Comment by mingxingtan** [comment]
I did a quick check and did not see issues. After you resolve the conflicts, we can merge this PR. @fsx950223

**Comment by fsx950223** [comment]
Fix #586, Fix #589

File: efficientdet/keras/train_lib.py [hunk 4]
```diff
@@ -183,6 +184,10 @@ def get_optimizer(params):
     optimizer = tf.keras.optimizers.Adam(learning_rate)
   else:
     raise ValueError('optimizers should be adam or sgd')
+  if params['moving_average_decay']:
+    # Only work on tfa-nightly
+    import tensorflow_addons as tfa
+    optimizer = tfa.optimizers.MovingAverage(optimizer, dynamic_decay=True)
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib.py:190]: Very nice! But I did not find "dynamic_decay" option on the doc: https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/MovingAverage

Looks like you should also set average_decay=params['moving_average_decay']?

Comment (fsx950223) [reply efficientdet/keras/train_lib.py:190]: Only work with addons nightly
https://github.com/tensorflow/addons/blob/a0bfe3f35f310bdfee53078bdee15e10af342f53/tensorflow_addons/optimizers/moving_average.py#L52
